core/src/main/scala/org/http4s/Method.scala

```diff
@@ -1,104 +1,115 @@
+  */
+sealed abstract case class Method private (name: String) extends Renderable {
+  final override def render(writer: Writer): writer.type = writer << name
 
-import Method.Semantics
+  /** Request methods are considered "safe" if their defined semantics are
+    * essentially read-only
+    *
+    * @see https://tools.ietf.org/html/rfc7231#section-4.2.1
+    */
+  def isSafe: Boolean
 
-/**
- * An HTTP method.
- *
- * @see [http://tools.ietf.org/html/rfc7231#section-4 RFC7321, Section 4]
- * @see [http://www.iana.org/assignments/http-methods/http-methods.xhtml IANA HTTP Method Registry]
- */
-sealed abstract case class Method private (name: String) extends Renderable with Semantics {
-  final override def render(writer: Writer): writer.type = writer << name
+  /** A request method is considered "idempotent" if the intended effect on the
+    * server of multiple identical requests with that method is the same as the
+    * effect for a single such request.
+    *
+    * @see https://tools.ietf.org/html/rfc7231#section-4.2.2
+    */
+  def isIdempotent: Boolean
+
+  /** Request methods can be defined as "cacheable" to indicate that responses to
+   *  them are allowed to be stored for future reuse
+   *
+   *  @see https://tools.ietf.org/html/rfc7231#section-4.2.3
+   */
+  def isCacheable: Boolean
 }
 
 object Method extends MethodInstances {
-  sealed trait Semantics {
-    def isIdempotent: Boolean
-    def isSafe: Boolean
-  }
-
-  object Semantics {
-    trait Default extends Semantics {
-      def isIdempotent: Boolean = false
-      def isSafe: Boolean = false
-    }
-    trait Idempotent extends Semantics {
-      def isIdempotent: Boolean = true
-      def isSafe: Boolean = false
-    }
-    trait Safe extends Semantics {
-      def isIdempotent: Boolean = true
-      def isSafe: Boolean = true
-    }
-  }
-
   // Type tags for a method allowing a body or not
   sealed trait PermitsBody extends Method
   sealed trait NoBody extends Method
 
+  private[http4s] class MethodImpl(name: String, val isIdempotent: Boolean = false, val isSafe: Boolean = false, val isCacheable: Boolean = false)
+      extends Method(name)
+
+  /** Resolves a method from a string. If the name is invalid, throws a
+    * [[ParseFailure]] */
   def fromString(s: String): ParseResult[Method] =
-    registry.getOrElse(s, Rfc2616BasicRules.token(s).bimap(
+    registry.getOrElse(s, HttpParser.token(s).bimap(
       e => ParseFailure("Invalid method", e.details),
-      new Method(_) with Semantics.Default
+      new MethodImpl(_)
     ))
 
-  import Semantics._
+  /** Resolves a method from a string. If the name is invalid, throws a
+    * [[ParseFailure]]. */
+  def unsafeFromString(s: String): Method =
+    fromString(s).fold(throw _, identity)
 
   // Lookups will usually be on fromString, so we store it wrapped in a \/-
-  private val registry = TrieMap[String, \/-[Method]]()
+  private var registry =
```

Comment:
As this is private, as is the register method. Do we expect this to be overridden and updated somewhere or could this instead be an Immutable Map?